Improve documentation and implementation of seismic anomalies post-processor

bobmyhill opened the pull request:
This PR addresses the problems with the seismic anomalies post-processor outlined in Issue #96 by:
* Adding n_slices as a user-defined parameter
* Improving the description of the post-processor, and increasing the number of comments
* Changing the processing of the lateral average calculation. The old code evaluated the average Vp and Vs in 50 depth slices, then took a 3-point simple running average of those values, and used that average for all cells within each depth range. The new code now evaluates the average Vp and Vs in n depth slices, then interpolates linearly between these values at the depths of each cell. This avoids systematic under- and overshooting at the edges of each depth slice.

gassmoeller commented on include/aspect/postprocess/visualization/seismic_anomalies.h:
```diff
@@ -61,6 +61,23 @@ namespace aspect
           execute () const;
+
+          /**
+                 * Declare the parameters this class takes through input files.
```
indent these lines please (unfortunately astyle does not fix comment lines).

gassmoeller commented on include/aspect/postprocess/visualization/seismic_anomalies.h:
```diff
@@ -61,6 +61,23 @@ namespace aspect
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int n_slices;
```
Document this parameter please.

gassmoeller commented on include/aspect/postprocess/visualization/seismic_anomalies.h:
```diff
@@ -93,6 +110,23 @@ namespace aspect
           execute () const;
+
+          /**
+                 * Declare the parameters this class takes through input files.
```
Indentation

gassmoeller commented on include/aspect/postprocess/visualization/seismic_anomalies.h:
```diff
@@ -93,6 +110,23 @@ namespace aspect
+          parse_parameters (ParameterHandler &prm);
+
+        private:
+          unsigned int n_slices;
```
Documentation please :smile:

gassmoeller commented on source/postprocess/visualization/seismic_anomalies.cc:
```diff
@@ -68,17 +40,24 @@ namespace aspect
+        std::vector<double> Vs_depth_average(n_slices);
+        this->get_lateral_averaging().get_Vs_averages(Vs_depth_average);
+
+        // Pad averages
```
Can you quickly explaine in the comment why you pad?

gassmoeller commented on source/postprocess/visualization/seismic_anomalies.cc:
```diff
@@ -116,22 +99,32 @@ namespace aspect
-              (*return_value.second)(cell_index) = (Vs - Vs_depth_average[idx])/Vs_depth_average[idx]*1e2;
+              // N.B. static_cast<int> always truncates the double (i.e. 1.9 -> 1)
+              // Here we subtract 0.5 to find the first slice with its center shallower than the cell,
+              // and add 1 to take padding into account
```
I do not see an add 1 in the code?

gassmoeller commented on source/postprocess/visualization/seismic_anomalies.cc:
```diff
@@ -116,22 +99,32 @@ namespace aspect
 
-              // compute the deviation from the average in per cent
-              (*return_value.second)(cell_index) = (Vs - Vs_depth_average[idx])/Vs_depth_average[idx]*1e2;
+              // N.B. static_cast<int> always truncates the double (i.e. 1.9 -> 1)
```
What does N.B. stand for?

gassmoeller commented on source/postprocess/visualization/seismic_anomalies.cc:
```diff
@@ -140,17 +133,24 @@ namespace aspect
+        std::vector<double> Vp_depth_average(n_slices);
+        this->get_lateral_averaging().get_Vp_averages(Vp_depth_average);
+
+        // Pad averages
```
same as above

gassmoeller commented on source/postprocess/visualization/seismic_anomalies.cc:
```diff
@@ -188,25 +192,121 @@ namespace aspect
 
-              // compute the deviation from the average in per cent
-              (*return_value.second)(cell_index) = (Vp - Vp_depth_average[idx])/Vp_depth_average[idx]*1e2;
+              // N.B. static_cast<int> always truncates the double (i.e. 1.9 -> 1)
```
same as above

gassmoeller commented:
/run-tests

bobmyhill commented:
Thanks @gassmoeller , I've addressed all your points. Hopefully this should work better now.

bobmyhill commented:
Well, this grew a little bit! I added a couple more comments regarding choice of values for the model parameters, ran astyle (oops, must remember this next time) and checked and updated the output. 

For the steinberger_seismic_velocities example, we get about a factor of three improvement in lateral velocity estimation. I suspect that this is largely down to removing the running average. In larger models, the performance should be even better.

bangerth commented on include/aspect/postprocess/visualization/seismic_anomalies.h:
```diff
@@ -93,6 +115,28 @@ namespace aspect
+
+          /**
+          * Declare the parameters this class takes through input files.
+          */
```
same here

bangerth commented on include/aspect/postprocess/visualization/seismic_anomalies.h:
```diff
@@ -93,6 +115,28 @@ namespace aspect
+           * Number of depth slices used to define average
+           * seismic compressional wave velocities from which anomalies
+           * are calculated.
+           **/
```
and here

bangerth commented on tests/steinberger_seismic_velocities.prm:
```diff
@@ -86,12 +86,18 @@ subsection Postprocess
     set Time between graphical output = 0
     set Number of zones = 10
- end
+  end
```
nice catch :-)

gassmoeller commented:
This PR now has conflicts with master, sorry for that. Ask @jdannberg if you need help resolving them, I think she is responsible for the problem :smile:. Otherwise ready to merge when the conflicts are fixed.

gassmoeller commented:
Please leave me a note when this is ready from your side.

bobmyhill commented:
This is ready from my side! :)
